client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs

```diff
@@ -14,8 +14,8 @@ use core::str::FromStr;
 
 use iroha_wasm::{data_model::prelude::*, DebugUnwrapExt, Execute};
 
-#[iroha_wasm::entrypoint]
-fn smartcontract_entry_point(_account_id: AccountId) {
+#[iroha_wasm::trigger_entrypoint]
+fn trigger_entrypoint(_account_id: AccountId, _event: Event) {
```

Comment:
another possibility would have been to let smartcontract ask wsv about the context of it's execution. Here this would mean to query which event triggered it's execution. The same could be done with `AccountId`. That's pull vs push architecture. I'm not strongly advocating one vs the other, I'm just thinking what would be easier for the writer to comprehend. Especially, if the number of args in the input will grow in the future and if they'll be rarely used

Reply:
Personally I think that we can do a bit of both. The number of parameters of this function can be different, and we can use the macros to generate the query of the WSV as if it were a parameter, and only add the event to the smartcontract if we need it (which we can tell from the macro by either not having the parameter at all, or having it unused).

Reply:
As a bonus you now have only one `entrypoint` macro

Reply:
So do you suggest to make entrypoint function parameter-less? And query all needed parameters from fn body? To me it sounds like a step away from strong-typing from user point of view. I think an *event* which have triggered smartcontract execution will be usefull pretty always

Reply:
This is relatively similar to how Rust handles CLI arguments. Most other languages accept `int argc, char** argv`, but Rust has a parameter-less entrypoint, so what we're doing wouldn't be surprising to seasoned Rust engineers.

Reply:
But we can also make a concession to `Event` as an argument. 

Most executables are parameter-less, so for them the `iroha_wasm::entrypoint` behaviour doesn't change. 

Some of the trigger executables need the `Event`, but most don't, which is why we didn't have it as part of the original implementation. So `iroha_wasm::trigger_entrypoint` could parse the extra argument, and simply be syntactic sugar around the query in the `fn` body. That way, if you don't need an event, you don't waste time moving it from the outers scope.

Reply:
> To me it sounds like a step away from strong-typing from user point of view.

I don' think this would affect typing

> Fortunately, the number of parameters of the entrypoint function can vary without changing the number of parameters in the generated function.

I would take this suggestion which leaves the smartcontract API pretty much unaffected. In this way user would never manually execute queries like `get_account_id` or `get_event`, instead the macro would guess the queries from the function signature. Also, there should be no need for `needless_pass_by_value` anymore